### PR TITLE
Add New-GcLogSink cmdlet

### PR DESCRIPTION
**Pull request by quoctruong**
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-powershell/421)

### CODE REVIEW

File: Google.PowerShell/Logging/GcLogCmdlets.cs [hunk 1]
```diff
@@ -347,17 +286,95 @@ public object GetDynamicParameters()
             return _dynamicParameters;
         }
 
+        public string SelectedResourceType
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogCmdlets.cs]: Comment?

File: Google.PowerShell/Logging/GcLogSinkCmdlets.cs [hunk 2]
```diff
@@ -93,4 +95,141 @@ protected override void ProcessRecord()
...
+        /// The project to check for log entries. If not set via PowerShell parameter processing, will
+        /// default to the Cloud SDK's DefaultProject property.
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false)]
+        [ConfigPropertyName(CloudSdkSettings.CommonProperties.Project)]
+        public string Project { get; set; }
+
+        [Parameter(Mandatory = false)]
+        public LogEntryVersionFormat? OutputVersionFormat { get; set; }
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogSinkCmdlets.cs]: These are all public and should have comments, especially if they are user visivble

File: Google.PowerShell/Logging/GcLogSinkCmdlets.cs [hunk 3]
```diff
@@ -93,4 +95,141 @@ protected override void ProcessRecord()
...
+                Name = SinkName
+            };
+
+            string permissionRequest = "";
+
+            switch (ParameterSetName)
+            {
+                case ParameterSetNames.GcsBucketDestination:
+                    logSink.Destination = $"storage.googleapis.com/{GcsBucketDestination}";
+                    permissionRequest = $"'Owner' permission to the bucket '{GcsBucketDestination}'.";
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogSinkCmdlets.cs]: Will these apis not return proper error messages?

File: Google.PowerShell/Logging/GcLogSinkCmdlets.cs [hunk 4]
```diff
@@ -93,4 +95,141 @@ protected override void ProcessRecord()
...
+            {
+                Name = SinkName
+            };
+
+            string permissionRequest = "";
+
+            switch (ParameterSetName)
+            {
+                case ParameterSetNames.GcsBucketDestination:
+                    logSink.Destination = $"storage.googleapis.com/{GcsBucketDestination}";
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogSinkCmdlets.cs]: Is there any pre checking we can do on the destination params?

File: Google.PowerShell/Logging/GcLogSinkCmdlets.cs [hunk 5]
```diff
@@ -93,4 +95,141 @@ protected override void ProcessRecord()
...
+
+            if (After.HasValue)
+            {
+                logSink.StartTime = XmlConvert.ToString(After.Value, XmlDateTimeSerializationMode.Local);
+            }
+
+            ProjectsResource.SinksResource.CreateRequest createRequest = Service.Projects.Sinks.Create(logSink, $"projects/{Project}");
+            if (NoUniqueWriterIdentity.IsPresent)
+            {
+                createRequest.UniqueWriterIdentity = !NoUniqueWriterIdentity.ToBool();
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogSinkCmdlets.cs]: consider ternary

File: Google.PowerShell/Logging/GcLogSinkCmdlets.cs [hunk 6]
```diff
@@ -93,4 +95,141 @@ protected override void ProcessRecord()
...
+            else
+            {
+                createRequest.UniqueWriterIdentity = true;
+            }
+
+            try
+            {
+                LogSink createdSink = createRequest.Execute();
+                WriteObject(createdSink);
+                Host.UI.WriteLine($"Please remember to grant '{createdSink?.WriterIdentity}' {permissionRequest}");
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogSinkCmdlets.cs]: Why is this needed? Can you add a comment?

File: Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1 [hunk 7]
```diff
@@ -72,3 +69,347 @@ Describe "Get-GcpsLogSink" {
...
 }
+
+Describe "New-GcpsLogSink" {
+    It "should work with -GcsBucketDestination" {
+        $r = Get-Random
+        $bucket = "gcloud-powershell-testing-logsink-bucket-$r"
+        $sinkName = "gcps-new-gclogsink-$r"
+        try {
+            New-GcLogSink $sinkName -GcsBucketDestination $bucket
+            Start-Sleep -Seconds 1
```
Comment (iantalarico) [inline Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1]: Instead of a sleep can you poll for the logsink?  I would avoid sleeps if possible

File: Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1 [hunk 8]
```diff
@@ -72,3 +69,347 @@ Describe "Get-GcpsLogSink" {
...
+
+            $createdSink = Get-GcLogSink -Sink $sinkName
```
Comment (iantalarico) [inline Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1]: Maybe a util function for checking the sink (or parts of it).  It looks like a lot of shared code.

File: Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1 [hunk 9]
```diff
@@ -72,3 +69,347 @@ Describe "Get-GcpsLogSink" {
...
+        finally {
+            gcloud beta logging sinks delete $sinkName --quiet 2>$null
+            gcloud beta logging sinks delete $sinkNameTwo --quiet 2>$null
+        }
+    }
+
+
+    It "should work with -Before and -After" {
+        $r = Get-Random
+        # These topics already has permission for cloud-logs to write to.
```
Comment (iantalarico) [inline Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1]: wording, maybe:
These topics already have permission to write to cloud-logs?

and below

File: Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1 [hunk 10]
```diff
@@ -72,3 +69,347 @@ Describe "Get-GcpsLogSink" {
...
+            New-GcLogSink $sinkName -PubSubTopicDestination $pubsubTopic -NoUniqueWriterIdentity
+            New-GcLogSink $sinkNameTwo -PubSubTopicDestination $pubsubTopic
+            Start-Sleep -Seconds 1
+
+            $createdSink = Get-GcLogSink -Sink $sinkName
+            $createdSink | Should Not BeNullOrEmpty
+            $createdSink.Name | Should BeExactly $sinkName
+            $createdSink.Destination | Should BeExactly "pubsub.googleapis.com/projects/$project/topics/$pubsubTopic"
+            $createdSink.OutputVersionFormat | Should BeExactly V2
+            $createdSink.WriterIdentity | Should BeExactly "serviceAccount:cloud-logs@system.gserviceaccount.com"
```
Comment (iantalarico) [inline Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1]: What is cloud-logs here? Comment?

Comment (iantalarico) [inline Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1]: Should this ("cloud-logs@system.gserviceaccount.com") be a const?

File: Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1 [hunk 11]
```diff
@@ -72,3 +69,347 @@ Describe "Get-GcpsLogSink" {
...
+            $createdSink.Filter | Should Not BeNullOrEmpty
+            $createdSink.StartTime | Should BeNullOrEmpty
+            $createdSink.EndTime | Should BeNullOrEmpty
+
+            # Write a log entry to the log, we should be able to get it from the subscription since it will be exported to the topic.
+            New-GcLogEntry -LogName $logName -TextPayload $textPayload
+            # Write a different entry to a different log (we should not get this).
+            New-GcLogEntry -LogName $secondLogName -TextPayload "You should not get this."
+            Start-Sleep -Seconds 20
```
Comment (iantalarico) [inline Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1]: Why do we need to sleep here and below?

File: Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1 (hunk 7 above)
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1]: Start-Sleep actually does not use Thread.Sleep underneath. It uses ManualResetEvent.WaitOne

**Comment by quoctruong** [comment]
@iantalarico PTAL

File: Google.PowerShell/Logging/GcLogSinkCmdlets.cs [hunk 12]
```diff
@@ -96,6 +96,58 @@ protected override void ProcessRecord()
...
     }
 
+    /// <summary>
+    /// <para type="synopsis">
+    /// Creates a new log sink.
+    /// </para>
+    /// <para type="description">
+    /// Creates a new log sink to export log entries. The sink will be created in the default project if -Project is not used.
+    /// Will raise an error if the sink already exists.
+    /// There are 3 possible destinations for the sink: Google Cloud Storage bucket, Google BigQuery dataset
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogSinkCmdlets.cs]: Do the destinations need to exist or will they be created if they don't?

File: Google.PowerShell/Logging/GcLogSinkCmdlets.cs [hunk 13]
```diff
@@ -133,25 +185,58 @@ public enum LogEntryVersionFormat
...
+        /// The log entry format to use for this sink's exported log entries. The v2 format is used by default.
+        /// The v1 format is deprecated and should be used only as part of a migration effort to v2.
+        /// </para>
+        /// </summary>
         [Parameter(Mandatory = false)]
         public LogEntryVersionFormat? OutputVersionFormat { get; set; }
 
+        /// <summary>
+        /// <para type="description">
+        /// The name of the sink to be created, unique within the project.
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogSinkCmdlets.cs]: [must be] unique within the project?

File: Google.PowerShell/Logging/GcLogSinkCmdlets.cs [hunk 14]
```diff
@@ -208,19 +293,14 @@ protected override void ProcessRecord()
...
-            {
-                createRequest.UniqueWriterIdentity = true;
-            }
+            createRequest.UniqueWriterIdentity = NoUniqueWriterIdentity.IsPresent ? !NoUniqueWriterIdentity.ToBool() : true;
 
             try
             {
                 LogSink createdSink = createRequest.Execute();
                 WriteObject(createdSink);
+                // We wants to let the user knows that they have to grant appropriate permission to the writer identity
```
Comment (iantalarico) [inline Google.PowerShell/Logging/GcLogSinkCmdlets.cs]: we want

File: Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1 [hunk 15]
```diff
@@ -70,7 +70,68 @@ Describe "Get-GcpsLogSink" {
     }
 }
 
-Describe "New-GcpsLogSink" {
+function Test-GcLogSink (
+    [string]$name,
```
Comment (iantalarico) [inline Google.PowerShell.IntegrationTests/Logging/Logging.GcLogSink.Tests.ps1]: Can't you do the asserts in this function? That would give better test failure messages

**Comment by quoctruong** [comment]
@iantalarico PTAL

**Comment by quoctruong** [comment]
Thanks!
